File: 문제풀이/온라인 저지/백준/# 07. Silver Ⅳ/# 01018. 체스판 다시 칠하기/java/Main.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int answer = 32;

        int[][] counter = new int[n + 1][m + 1];

        String line;
        boolean flag;
        boolean current;
        int sum;
        int i, j;
        for (i = 1; i <= n; i++) {
            line = br.readLine();
            flag = i % 2 != 0;
            for (j = 1; j <= m; j++) {
                current = line.charAt(j - 1) == 'W';
                counter[i][j] = counter[i - 1][j] + counter[i][j - 1] - counter[i - 1][j - 1];
                if (flag != current) {
                    counter[i][j] += 1;
                }
                flag = !flag;

                if (i >= 8 && j >= 8) {
                    sum = counter[i][j] - counter[i - 8][j] - counter[i][j - 8] + counter[i - 8][j - 8];
                    if (sum > 32) sum = 64 - sum;
                    if (answer > sum) answer = sum;
                }
            }
        }
        System.out.print(answer);
    }

}

File: 문제풀이/온라인 저지/백준/# 07. Silver Ⅳ/# 01018. 체스판 다시 칠하기/kotlin/Main.kt
fun main() {
    val n = i()
    val m = i()
    val board = Array(n + 1) { IntArray(m + 1) }
    var flag: Boolean
    var sum: Int
    var answer = 64
    for (i in 1..n) {
        flag = i % 2 == 1
        for (j in 1..m) {
            board[i][j] =
                board[i - 1][j] + board[i][j - 1] - board[i - 1][j - 1] + if ((r() == 'B'.toByte()) == flag) 0 else 1
            flag = !flag

            if (i >= 8 && j >= 8) {
                sum = board[i][j] - board[i - 8][j] - board[i][j - 8] + board[i - 8][j - 8]
                if (sum > 32) sum = 64 - sum
                if (answer > sum) answer = sum
            }
        }
        r()
    }
    println(answer)
}


private const val S = 65536
private val iS = java.io.DataInputStream(System.`in`)
private val b = ByteArray(S)
private var c = 0
private var l = 0

private fun i(): Int {
    var v = 0
    var c = r()
    do v = v * 10 + c - 48 while (r().also { c = it } > 47)
    return v
}

private fun r(): Byte {
    if (c == l) {
        l = iS.read(b, 0.also { c = it }, S)
        if (l == -1) b[0] = -1
    }
    return b[c++]
}